### ISSUE TITLE
[Inductor][SM100] ptxas VIMNMX3 fusion drops the negated integer-abs operand

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#190973]
## Summary

This one was a wild ride :)

### Agent Report

On GB200 (SM100), optimized ptxas miscompiles a valid integer absolute-value sequence in an Inductor-generated Triton FlexAttention kernel. This corrupts a dynamic NATTEN mask: affected rows count 260 live keys instead of the expected `13 * 13 = 169`.

The issue is reduced from [`pytorch-labs/attention-gym::test_natten_masks`](https://github.com/pytorch-labs/attention-gym/blob/main/test/test_natten.py), which compares semantically equivalent row-major, tiled, and Morton layouts, including forward output, dQ, dK, and dV.

A draft Triton fix is available in:

- https://github.com/triton-lang/triton/pull/11032

### Original Attention Gym test

On an affected SM100 PyTorch/Triton installation:

```bash
git clone https://github.com/pytorch-labs/attention-gym.git
cd attention-gym
pip install -e ".[dev]"
TRITON_ALWAYS_COMPILE=1 TORCHINDUCTOR_CACHE_DIR=/tmp/natten-inductor TRITON_CACHE_DIR=/tmp/natten-triton pytest -q test/test_natten.py::test_natten_masks -vv --tb=short
```

The unmodified compiler fails a forward comparison between equivalent layouts. ptxas O0 makes the complete forward/backward test pass:

```bash
PTXAS_OPTIONS="--opt-level=0" TRITON_ALWAYS_COMPILE=1 pytest -q test/test_natten.py::test_natten_masks
```

### Trimmed Python/Triton reproducer

The reduced kernel removes FlexAttention and directly counts its integer neighborhood mask while keeping a zero-valued TCGen5 dot live:

Gist; https://gist.github.com/4ced47465f083119405a28e6a5833872

```bash
git clone https://gist.github.com/4ced47465f083119405a28e6a5833872.git sm100-ptxas-repro
cd sm100-ptxas-repro
TRITON_ALWAYS_COMPILE=1 TRITON_CACHE_DIR=/tmp/sm100-abs-repro python repro.py
```

Expected result before the workaround:

```text
expected key count: 169
maximum key-count error: 819.0
AssertionError
```

The same script reports zero error with [Triton PR #11032](https://github.com/triton-lang/triton/pull/11032).

### Standalone C++/PTX reproducer

The same gist contains the original 261 KB valid PTX, a CUDA Driver API launcher, and a script that compiles the PTX at O3 and O0. This path requires no Python, PyTorch, Triton, or Inductor at runtime:

```bash
git clone https://gist.github.com/4ced47465f083119405a28e6a5833872.git sm100-ptxas-repro
cd sm100-ptxas-repro
PTXAS=/path/to/ptxas-blackwell CUDA_HOME=/usr/local/cuda-13.1 bash run.sh
```

Verified on GB200 with every available ptxas version. The same PTX instruction stream was used with the `.version` declaration adjusted for each compiler:

| ptxas | PTX ISA | Optimized result | `--opt-level=0` result |
|---|---:|---:|---:|
| CUDA 12.8 (`V12.8.93`) | 8.7 | error 819; 20 / 128 rows wrong | error 0; 0 / 128 rows wrong |
| CUDA 13.0 (`V13.0.88`) | 9.0 | error 819; 20 / 128 rows wrong | error 0; 0 / 128 rows wrong |
| CUDA 13.1 (`V13.1.115`) | 9.1 | error 819; 20 / 128 rows wrong | error 0; 0 / 128 rows wrong |
| bundled 13.3 `ptxas-blackwell` (`V13.3.33`) | 9.3 | error 819; 20 / 128 rows wrong | error 0; 0 / 128 rows wrong |

## Root cause

LLVM emits valid PTX equivalent to:

```ptx
neg.s32 neg_x, x;
max.s32 abs_x, x, neg_x;
max.s32 distance, abs_x, abs_y;
```

ptxas O1-O3 incorrectly fuse one instance into:

```text
VIMNMX3 R167, R8, R8, R167, !PT
```

The negated operand has been coalesced with `R8`. For `x = 18, 17, ..., 0, -1, ..., -13` and `abs_y = 6`, cuda-gdb shows the result become:

```text
18, 17, ..., 7, 6, 6, ..., 6
```

This is `max(x, 6)`, not `max(abs(x), 6)`, so 20 lanes remain live instead of 13. The subsequent predicate, reduction, shared-memory transpose, accumulation, and store preserve that already-wrong value correctly.

The Triton PR canonicalizes emitted SM100+ PTX from the `neg.s32` + matching `max.s32` idiom to `abs.s32`, avoiding the malformed fusion while preserving the original negated register definition.

## Validation

With normal optimized ptxas and no optimization-disabling environment variables:

```text
python/test/unit/language/test_core.py::test_integer_abs_ptxas_fusion
before fix: maximum count error 819
with fix:   passed

attention-gym/test/test_natten.py::test_natten_masks
with fix:   1 passed in 21.77s
```

The end-to-end test passes all row-major, tiled, and Morton forward, dQ, dK, and dV comparisons. `PTXAS_OPTIONS` and `DISABLE_PTXAS_OPT` were unset.

<details>
<summary>How the issue was narrowed down</summary>

### 1. Excluded NATTEN mask semantics

The original failure compared row-major, tiled, and Morton token layouts. One million sampled `(q_idx, kv_idx)` pairs produced no logical-mask differences after mapping each layout back to row-major coordinates. A direct reference also showed that the generated Triton result, rather than the Python mask, was wrong.

### 2. Turned attention into an exact key-count test

Setting `Q=K=0` makes every unmasked score zero. The softmax population can then be read exactly: each query should see `13 * 13 = 169` keys. Affected rows instead reported 260, corresponding to seven extra columns in each of 13 image rows:

```text
13 * (13 + 7) = 260
```

The TCGen5 dot output remained exactly zero.

### 3. Removed FlexAttention and softmax from the reproducer

A standalone Triton kernel kept the zero-valued `tl.dot` live through an independent checksum but directly summed the boolean mask. It produced:

```text
max mask-count error = 91
max dot checksum = 0
```

This proved that the dynamic predicate itself was corrupted; the result was not caused by online-softmax arithmetic.

### 4. Located the compiler boundary

The same SLP-enabled PTX is correct with ptxas O0 and wrong with O1, O2, or O3. TTIR, TTGIR, LLVM IR, and PTX are byte-identical between the ptxas O0/O3 runs; only cubin/SASS changes. The failure reproduces with ptxas from CUDA 12.8, 13.0, 13.1, and bundled 13.3.

Skipping LLVM's `SLPVectorizerPass` changed the PTX shape enough to fix forward, but dQ remained wrong. That showed SLP was an exposing code-shape trigger rather than the semantic miscompile.

### 5. Excluded common runtime hazards

The reduced kernel remains wrong under every Compute Sanitizer checker, with no findings:

```text
memcheck:  0 errors
initcheck: 0 errors
racecheck: 0 hazards, errors, or warnings
synccheck: 0 errors
```

### 6. Traced the bad value backward with cuda-gdb

Immediately before the final store for output row 114:

```text
R2 = 260.0  # mask count
R3 = 0.0    # independent dot checksum
```

Each active 64-key half-tile had already reduced to 20 instead of 13. Inactive halves were zero, and the accumulator correctly added thirteen contributions of 20. Shared-memory transfer and the global store also preserved the already-wrong value.

Tracing the predicate further backward reached the first incorrect instruction. Before it, cuda-gdb observed:

```text
x     = 18, 17, ..., 1, 0, -1, ..., -13
abs_y = 6
```

Valid PTX computes `max(abs(x), abs_y)` using `neg.s32`, `max.s32`, and another `max.s32`. Optimized ptxas instead emitted:

```text
VIMNMX3 R167, R8, R8, R167, !PT
```

The two sources that should represent `x` and `-x` are both `R8`. Its runtime output is `18, 17, ..., 7, 6, 6, ...`, which is `max(x, 6)` rather than `max(abs(x), 6)`. This leaves 20 lanes live instead of 13.

A tiny isolated PTX abs/max expression compiles to a correct `VIMNMX3` with distinct `x` and `-x` operands, so the instruction itself is valid. The bug is exposed by optimization/register allocation in the larger kernel.

### 7. Confirmed the fix with controlled PTX A/Bs

Rewriting only the steady-loop absolute value changed row 114 from 260 to 176, fixing twelve iterations. Rewriting its separately generated epilogue copy changed 176 to the expected 169. Canonicalizing all matching integer-abs idioms to `abs.s32` fixed the reduced kernel and the complete row-major/tiled/Morton forward, dQ, dK, and dV test.

The focused SM100 regression and workaround are in [Triton PR #11032](https://github.com/triton-lang/triton/pull/11032).

</details>

<details>
<summary>Environment</summary>

- PyTorch: `2.14.0.dev20260723+cu132`
- PyTorch commit: `77b8883fd9e1659e5a540711528ee34c0139629f`
- Triton release package: `3.8.0`
- Triton main used for the fix: `947a8528bc64ddf6667762432cc116de5b797f9c`
- Python: `3.13.13` (Clang `22.1.3`)
- GPU: NVIDIA GB200, compute capability `(10, 0)`
- NVIDIA driver: `580.126.20`
- PyTorch CUDA build: `13.2`
- CUDA runtime: `13.1.115`
- Bundled `ptxas-blackwell`: CUDA `13.3`, `V13.3.33`, build `37862127`
- Also reproduced with ptxas from CUDA `12.8`, `13.0`, and `13.1`
- OS: CentOS Stream 9, aarch64
- GCC: `11.5.0`
- glibc: `2.34`

</details>

cc @ezyang @gchanan @kadeng @msaroufim @ptrblck @eqy @tinglvv @nWEIdia @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo @bertmaher @int3 @davidberard98 @nmacchioni @embg @peterbell10 @iupaikov-amd @njriasan

**Comment by eqy** [comment]
Couldn't reproduce this immediately locally but my closest setup was Triton 3.7.1 + 13.3 `ptxas` on GB200

Opened an issue with `ptxas` team to track

**Comment by drisspg** [comment]
Confirmed that on the V13.4.46 ptxas preview we this bug is fixed :)

**Comment by drisspg** [comment]
cc @warrendeng Any update here? did we figure out what ptxas we are going to ship with?

**Comment by atalman** [comment]
HI @drisspg for 2.14.0 release  we unable to include V13.4.46 ptxas preview since its not officially released yet.

**Comment by nWEIdia** [comment]
> HI [@drisspg](https://github.com/drisspg) for 2.14.0 release we unable to include V13.4.46 ptxas preview since its not officially released yet.

In nightly PyTorch (CUDA 13.4 only), this ptxas is packed with Torch wheel and can be accessed via  PYTORCH_INSTALL_DIRECTORY/bin/ptxas

**Comment by eellison** [comment]
since this is related to release packaging, assigning to @atalman .
